source/ncengine/graphics2/frontend/subsystem/InstanceCache.h

```diff
@@ -0,0 +1,198 @@
+#pragma once
+
+#include "graphics2/ShaderTypes.h"
+#include "graphics2/frontend/subsystem/MeshRendererRenderState.h"
+
+#include "ncengine/asset/AssetViews.h"
+#include "ncengine/utility/SparseMap.h"
+
+#include <limits>
+#include <ranges>
+#include <vector>
+
+namespace nc::graphics
+{
+constexpr auto NullBatchIndex = std::numeric_limits<uint32_t>::max();
+
+// A batch is uniquely identified by a mesh and set of material passes.
+struct BatchKey
+{
+    MaterialPasses passes = NullBatchIndex;
+    uint64_t meshId = std::numeric_limits<uint64_t>::max();
+
+    friend auto operator==(const BatchKey& lhs, const BatchKey& rhs) -> bool
+    {
+        return lhs.passes == rhs.passes && lhs.meshId == rhs.meshId;
+    }
+};
+
+// Batch state describing a contiguous region in the instance buffer as well as draw attributes.
+struct BatchRegion
+{
+    BatchKey key = BatchKey{};
+    Batch batch;
+    uint32_t capacity = 0;
+};
+
+// A new batch waiting to be added to the instance buffer.
+struct StagedBatchRegion
+{
+    MaterialPasses passes = MaterialPasses{};
+    asset::MeshView mesh = asset::MeshView{};
+};
+
+// InstanceData for a MeshRenderer waiting to added to a batch.
+struct StagedInstance
+{
+    uint32_t entityId = NullBatchIndex;
+    InstanceData instanceData = InstanceData{};
+};
+
+// Per-batch staging area.
+struct StagedBatchInstances
+{
+    BatchKey key = BatchKey{};
+    std::vector<StagedInstance> instances = {};
+};
+
+// MeshRenderer instance waiting to be removed from a batch.
+struct StagedRemoval
+{
+    BatchKey key = BatchKey{};
+    uint32_t entityId = NullBatchIndex;
+};
+
+// InstanceCache interface for adding, removing, and updating MeshRenderer instances. Events
+// originating from game logic/API should only be routed to here to avoid race conditions
+// with the actual InstanceCache buffer.
+class InstanceCacheStaging
+{
+    public:
+        // API-Facing Functions
+        void AddInstance(uint32_t entityId,
+                         MaterialPasses passes,
+                         const asset::MeshView& mesh,
+                         const InstanceData& instanceData);
+
+        void RemoveInstance(uint32_t entityId,
```

Comment:
I don't have a problem with it just curious: what's the reasoning behind leaving the batches around when they've been removed of all their members?

Reply:
Its to minimize work. Thinking being that if a batch existed once, it'll probably get used again. By leaving it around, those spaces on the gpu are marked for reuse later, and we don't have to send anything over the wire on removal. If/when a new object is added to the batch in the future, there's already a space reserved for it.

They are removed on scene change though, because a batch could be based on a mesh that's specific to a scene.